gapis/trace/android/adreno/validate.go

```diff
@@ -34,17 +44,77 @@ var (
 		{37, "% Time Shading Fragments", validate.And(validate.IsNumber, validate.CheckLargerThanZero)},
 		{38, "% Time Shading Vertices", validate.And(validate.IsNumber, validate.CheckLargerThanZero)},
 		{39, "% Time Compute", validate.And(validate.IsNumber, validate.CheckEqualTo(0.0))},
-		{58, "Avg Memory Latency Cycles", validate.And(validate.IsNumber, validate.CheckLargerThanZero)},
 	}
 )
 
 type AdrenoValidator struct {
 }
 
+func (v *AdrenoValidator) validateRenderStage(ctx context.Context, processor *perfetto.Processor) error {
+	tIds, err := validate.GetRenderStageTrackIDs(ctx, processor)
+	if err != nil {
+		return err
+	}
+	for _, tId := range tIds {
+		queryResult, err := processor.Query(fmt.Sprintf(renderStageSlicesQuery, tId))
+		if err != nil || queryResult.GetNumRecords() <= 0 {
+			return log.Errf(ctx, err, "Failed to query with %v", fmt.Sprintf(renderStageSlicesQuery, tId))
+		}
+		columns := queryResult.GetColumns()
+		names := columns[0].GetStringValues()
+
+		// Skip slices until we hit the first 'Surface' slice.
+		skipNum := -1
+		hasSurfaceSlice := false
+		hasRenderSlice := false
+		for i, name := range names {
+			if name == "Surface" {
+				hasSurfaceSlice = true
+			}
+			if name == "Render" {
+				hasRenderSlice = true
+			}
+			if skipNum == -1 && name == "Surface" {
+				skipNum = i
+			}
+		}
+		if !hasSurfaceSlice {
+			return log.Errf(ctx, err, "Render stage verification failed: No Surface slice found")
+		}
+		if !hasRenderSlice {
+			return log.Errf(ctx, err, "Render stage verification failed: No Render slice found")
+		}
+		names = names[skipNum:]
+		depths := columns[1].GetLongValues()[skipNum:]
+		parentStackId := columns[2].GetLongValues()[skipNum:]
+
+		for i := 0; i < len(names); i++ {
```

Comment:
maybe just `for i := skipNum; ...` then you don't need to slice things, but either is fine.